File: kotlin/lib/library/src/androidTest/java/com/xplpc/library/LocalClientTest.kt
package com.xplpc.library

import android.content.Context
import android.os.BatteryManager
import androidx.test.ext.junit.runners.AndroidJUnit4
import androidx.test.platform.app.InstrumentationRegistry
import com.xplpc.client.LocalClient
import com.xplpc.core.Config
import com.xplpc.core.XPLPC
import com.xplpc.data.MappingData
import com.xplpc.map.MappingItem
import com.xplpc.message.Message
import com.xplpc.message.Param
import com.xplpc.message.Request
import com.xplpc.message.Response
import com.xplpc.serializer.JsonSerializer
import kotlinx.coroutines.runBlocking
import org.junit.Assert.assertEquals
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith

@RunWith(AndroidJUnit4::class)
class LocalClientTest {
    companion object {
        fun batteryLevel(data: Message): Response {
            val appContext = InstrumentationRegistry.getInstrumentation().targetContext
            val bm = appContext.getSystemService(Context.BATTERY_SERVICE) as BatteryManager
            val level = bm.getIntProperty(BatteryManager.BATTERY_PROPERTY_CAPACITY)

            val suffix = data.value<String>("suffix")

            return if (suffix != null) {
                Response("$level$suffix")
            } else {
                Response("$level")
            }
        }
    }

    @Before
    fun setUp() {
        XPLPC.initialize(
            Config(JsonSerializer()),
        )
    }

    @Test
    fun batteryLevel() {
        MappingData.add(
            MappingItem(
                "platform.battery.level",
                LocalClientTest::batteryLevel
            )
        )

        val request = Request(
            "platform.battery.level",
            Param("suffix", "%")
        )

        val level = LocalClient.call<String>(request) ?: "ERROR"

        assertEquals("100%", level)
    }

    @Test
    fun batteryLevelAsync() {
        MappingData.add(
            MappingItem(
                "platform.battery.level",
                LocalClientTest::batteryLevel
            )
        )

        val request = Request(
            "platform.battery.level",
            Param("suffix", "%")
        )

        runBlocking {
            val level = LocalClient.callAsync<String>(request).await() ?: "ERROR"
            assertEquals("100%", level)
        }
    }
}

File: swift/tests/Sources/LocalClientTest.swift
import XCTest

@testable import xplpc

final class LocalClientTest: XCTestCase {
    override func setUpWithError() throws {
        // Put setup code here. This method is called before the invocation of each test method in the class.
        XPLPC.shared.initialize(
            config: Config(serializer: JsonSerializer())
        )
    }

    override func tearDownWithError() throws {
        // Put teardown code here. This method is called after the invocation of each test method in the class.
    }

    func batteryLevel(message: Message, r: Response) {
        let level = 100

        let suffix: String = message.get("suffix") ?? ""

        if suffix.isEmpty {
            r("\(level)")
        } else {
            r("\(level)\(suffix)")
        }
    }

    func testBatteryLevel() throws {
        MappingList.shared.add(name: "platform.battery.level", item: MappingItem(target: batteryLevel))

        let request = Request("platform.battery.level", Param("suffix", "%"))

        LocalClient.call(request) { response in
            XCTAssertEqual("100%", response)
        }
    }

    func testBatteryLevelAsync() throws {
        MappingList.shared.add(name: "platform.battery.level", item: MappingItem(target: batteryLevel))

        let request = Request("platform.battery.level", Param("suffix", "%"))

        DispatchQueue.global(qos: .background).async {
            LocalClient.call(request) { response in
                XCTAssertEqual("100%", response)
            }
        }
    }

    func testBatteryLevelInvalidCast() throws {
        MappingList.shared.add(name: "platform.battery.level", item: MappingItem(target: batteryLevel))

        let request = Request("platform.battery.level", Param("suffix", "%"))

        LocalClient.call(request) { (response: Bool?) in
            XCTAssertEqual(nil, response)
        }
    }
}
